### PR TITLE
Avoid unnecessarily resizing dictionary while calculating project closure

### LINKED ISSUE #10976: Avoid unnecessarily resizing dictionary while calculating project closure
Thanks to @davkean for identifying a performance improvement in NuGet.

`ToDictionary does not let you specify an underlying starting capacity. This causes it to inherit the default size and unnecessarily resize itself multiple times in a large solution. Added a version of ToDictionary that lets you specify a default.`

https://github.com/davkean/NuGet.Client/blob/e97d556a83750bb4f8017e3640bc699d04832657/src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs#L191-L192

This was allocating 0.5% (150 MB) opening a 1000 project solution.

https://github.com/NuGet/NuGet.Client/pull/4129#issue-679612890

### PR DESCRIPTION
**Pull request by kartheekp-ms**
## Bug

Fixes: https://github.com/NuGet/Home/issues/10976

Regression? Last working version:

## Description

Creating this PR as per https://github.com/NuGet/NuGet.Client/pull/4129#discussion_r661034629 conversation. More details about this performance improvement can be found in the above-mentioned PR.

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [x] N/A  No functional change to the product.

- **Documentation**
  - [x] N/A

### CODE REVIEW

File: src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs
Comment (erdembayar) [inline src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs:166]: It looks previous implementation is check OS specific uniqueness of dictionary key. Does `_projects` satisfy this requirement for different OS like Win, MAC, Linux etc?

Comment (kartheekp-ms) [reply src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs:166]: My understanding is `PackageSpec.RestoreMetadata.ProjectUniqueName` is used as a key when adding items to the dictionary. Hence the casing should be the same.  Unless there is a gap in the test coverage for this code path, CI produced a green build for this change.

https://github.com/NuGet/NuGet.Client/blob/b102b7587f0286d83dad8ee357caded56bd0ab17/src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs#L216-L226

**Review by nkolev92 — APPROVED** [review]
The current file has unnecessary usings, given that you're editing it, consider cleaning them up.

File: src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs
Comment (kartheekp-ms) [inline src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs:311]: This method has no references in the code.

**Comment by kartheekp-ms** [comment]
@nkolev92 - I removed unnecessary `using` statements and applied few IDE suggestions in the latest commit.

File: src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs
Comment (nkolev92) [inline src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs:229]: Given that this is a suggestion and syntactic sugar, I personally don't think we should too aggressive in makes changes like this since they heavily affected the `blame-ability` of a file. 

I'm not asking you to revert it, but in general, I think we should be less aggressive with making large changes

**Review by nkolev92 — APPROVED** [review]

**Comment by kartheekp-ms** [comment]
Waiting for second approval on this PR. Thanks

File: src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs
Comment (zivkan) [reply src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs:166]: > My understanding is PackageSpec.RestoreMetadata.ProjectUniqueName is used as a key when adding items to the dictionary. Hence the casing should be the same.

The fact that the project unique name is used does not, by itself, ensure that the behaviour is the same. Imagine two different projects, both of which have a `ProjectReference` to a third project, but use different casing in their `Include` path. On Linux, this will fail* regardless, but on Windows it could succeed* if all the code that uses the path knows to treat it as case insensitive.

*On Linux, Windows, and Mac, file sysstem can be configured to be either case sensitive or insensitive. It's just the deaults on the various operating systems that are different. Applications like NuGet that try to treat it always being the same are going to have bugs when the customer changes their filesystems' settings.


Having said that if @erdembayar checked the definition/initiaization of `_projects`, he'd see that `PathUtility.GetStringComparerBasedOnOS()` was also used.  This is the reason why the behaviour isn't changing with this code change.

File: src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs
Comment (zivkan) [reply src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs:229]: In general, I'm more lenient than Nikolche about code cleanup, but the first commit in this branch had 7 lines changed, whereas the second had 125 lines changed.  This makes it difficult to see what the actual product changes are. But thank you for doing it in separate commits. It helped me a lot in reviewing this PR.

**Review by zivkan — APPROVED** [review]

File: src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs
Comment (kartheekp-ms) [reply src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs:166]: Thanks, Andy, for additional information. As Andy pointed out `PathUtility.GetStringComparerBasedOnOS()` invoked when `_projects` was defined. 

https://github.com/NuGet/NuGet.Client/blob/a5d5e3236e86f6c3c09097b723ce2b52852a1752/src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs#L22

**Pull request merged**
